backends/ebpf/runtime/ebpf_runtime.c

```diff
@@ -14,89 +14,164 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-/* The runtime emulates a loaded eBPF program and provides an interface to feed
-   packets to the emulated eBPF filter. */
-
-
-
-#include <string.h>
-#include <libgen.h>
-#define PCAP_DONT_INCLUDE_PCAP_BPF_H
-#include <pcap/pcap.h>
+/* The runtime emulates a loaded eBPF program. It parses a set of given
+  pcap files and executes a C program that processes packets extracted from
+  the capture files. The C program takes a packet structure (commonly an skbuf)
+  and returns an action value for each packet.
+ */
+
+#include <unistd.h>     // getopt()
+#include <ctype.h>      // isprint()
 #include "test.h"
+#ifdef CONTROL_PLANE
 #include "control.h"
-#include <unistd.h>
-#include <ctype.h>
+#endif
+#include "pcap_util.h"
+
 #define FILE_NAME_MAX 256
+#define PCAPIN  "_in.pcap"
+#define PCAPOUT "_out.pcap"
+#define DELIM   '_'
 
 static int debug = 0;
 
 void usage(char *name) {
-    fprintf(stderr, "This program parses a pcap file, "
-            "feeds the individual packets into a filter function, "
+    fprintf(stderr, "This program expects a pcap file pattern, "
+            "extracts all the packets out of the matched files"
+            "in the order given by the packet time,"
+            "then feeds the individual packets into a filter function, "
             "and returns the output.\n");
-    fprintf(stderr, "Usage: %s [-d] -f file.pcap\n", name);
+    fprintf(stderr, "Usage: %s [-d] -f file.pcap [-n num_files]\n", name);
     fprintf(stderr, "Options:\n");
     fprintf(stderr, "\t-d: Turn on debug messages\n");
     fprintf(stderr, "\t-f: The input pcap file\n");
+    fprintf(stderr, "\t-n: Specifies the number of input files\n");
     exit(EXIT_FAILURE);
 }
 
-static int process_packets(pcap_t *in_handle, pcap_dumper_t *out_handle) {
-    struct pcap_pkthdr *pcap_hdr;
-    const unsigned char *packet;
-    int ret;
-    while ((ret = pcap_next_ex(in_handle, &pcap_hdr, &packet)) == 1) {
-        /* Parse each packet in the file and check the result */
+static pcap_list_t *get_packets(const char *pcap_base, uint16_t num_files) {
+    pcap_list_array_t *tmp_list_array = allocate_pkt_list_array();
+    /* Retrieve a list for each file and append it to the temporary array */
+    for (uint16_t i = 0; i < num_files; i++) {
+        char pcap_in_name[FILE_NAME_MAX];
+        snprintf(pcap_in_name, FILE_NAME_MAX, "%s%d%s", pcap_base, i, PCAPIN);
+        if (debug)
+            printf("Processing input file: %s\n", pcap_in_name);
+        pcap_list_t *pkt_list = read_pkts_from_pcap(pcap_in_name, i);
+        tmp_list_array = append_list(tmp_list_array, pkt_list);
```

Comment:
Yes, because the pointer might get updated by realloc. I could also enforce passing a double pointer instead. The list array is deleted in merge_lists(). I will rename that function to make it more explicit.